### ISSUE TITLE
Implement a filter boolean variable for deleted/hidden messages. 

### ISSUE THREAD

**Issue by XargsUK** [Chatterino/chatterino2#2773]
### Issue

Currently, messages which are deleted through moderation or hidden through R9K have a global Show/Hide option (Show deleted messages). This can lead to users having to toggle the Checkbox to see messages which have been deleted. 

### Enhancement

Implement a filter boolean variable for deleted/hidden messages. 

### Pre-Implementation Scenario

- User 1 has a number of Splits on one Tab all showing the same channel. 
- User 1 has different filters on different Splits. 
- User 1 has _Show deleted messages_ unchecked. 
- User 2 mentions User 1 in the channel. The message is removed (due to length, moderation or R9K). 
- User 1 enabled _Show deleted messages_ to see the mention. The feed now displays all the R9K hidden messages. 

### Post-Implementation Scenario

- User 1 has a number of Splits on one Tab all showing the same channel. 
- User 1 has different filters on different Splits, some with a deleted/hidden message boolean set to true. 
- User 2 mentions User 1 in the channel. The message is removed (due to length, moderation or R9K). 
- User 1 can see the hidden/deleted message in a Split with a Filter Boolean of hidden/deleted message set to true.

**Comment by apa420** [comment]
I don't understand at all what you're trying to achieve.
If you want to see all messages just leave the options off?
If I have deleted messages set as hidden I don't want to see them. So what is the issue?

**Comment by XargsUK** [comment]
So with this logic, why allow users different filters for each Split. If you want to see things based on xyz variable, just have a global setting to turn it on. If you don't want to see it, turn it off. The brilliance behind the filters is that it ultimately provides granular control for each split. For example, if you wanted to only see people in chat with a specific badge, you would likely not want this filter applied to every tab and split across the whole application, but rather a specific split. 

The point here is being able to show/hide hidden messages on a per-split basis. In my own personal use-case, I have one channel open with 4 splits, all displaying the same chat feed. Each split has different filters applied, 1 for the main chat, one for specific peoples messages, one for messages containing a specific terms etc. I also have R9K enabled, as the chat feed is extremely active. If a user @ mentions me and it breaches a moderation character limit, it is deleted but a notification is sent. I then have to disable Show Deleted Messages. When it's disabled, I can then see the message mentioning me, but then R9Ks hidden messages show in the feed (though greyed out), which defeats the purpose of trying to lower the busyness of the feed. This is one specific use-case, though I have more if you'd like?

**Comment by Mm2PL** [comment]
I don't think this will work with the filters system. IIRC filters don't work on older messages, so this variable probably wouldn't work.

**Comment by apa420** [comment]
So basically the `R9K` option and `hide deleted messages` should be part of filters instead?

**Comment by XargsUK** [comment]
Correct @apa420. Also @Mm2PL your reply makes sense. Once passed through a filter, it wouldn't go through the filter again. However, with twitch chat specifically, the CLEARMSG and CLEARCHAT commands are in fact processed by Chatterino, which is how is marks them as deleted. Having a filter enabled for these two commands would allow you to trigger a reprocessing of the messages. 

E.g. User ronni messages HeyGuys. Moderation bot in twitch deletes it (`@login=ronni;target-msg-id=abc-123-def :tmi.twitch.tv CLEARMSG #dallas :HeyGuys`). Chatterino will hide the message if the Show Deleted Messages option is on, and I'm assuming this is based on this exact command. This could easily retrigger the message to be reprocessed by the filters. 

@apa420 Those two options, yeah. _Implement a filter boolean variable for deleted/hidden messages._

**Comment by XargsUK** [comment]
https://github.com/Chatterino/chatterino2/blob/master/src/providers/twitch/TwitchIrcServer.cpp

```
    else if (command == "CLEARCHAT")
    {
        handler.handleClearChatMessage(message);
    }
    else if (command == "CLEARMSG")
    {
        handler.handleClearMessageMessage(message);
    }
```

**Comment by apa420** [comment]
The reason I wrote it like that is because the `R9K` and `hide deleted messages` hide it from the actual channel and filters are only filters on top of the actual channel.

**Comment by XargsUK** [comment]
> The reason I wrote it like that is because the `R9K` and `hide deleted messages` hide it from the actual channel and filters are only filters on top of the actual channel.
> As far as I understand how it works.

Yeah, it makes sense. My proposal would be to change the handler.handleClearChatMessage(message); function to instead be an identifier so that they can be filtered. At the moment, when handler.handleClearChatMessage(message) is triggered, it is removed from the split. Instead of this, it could update a boolean identifier and refilter. 

Apologises as well, I might be really misunderstanding the way to which the application functions here, but really appreciate the input :)

**Comment by fourtf** [comment]
I removed stuff from the issue that should be left out as described by the issue template. It caused the comments to immediately derail into pointless technicallities.
